### PR TITLE
Capture CCPA consent data through Ophan

### PR DESCRIPTION
**Pull request by shtukas**
## Context

So far, we have recorded consent information from the client side to the datalake through Ophan using the Ophan Component Event to carry the data. For context, using the component event to perform that task is not ideal, but at least it works for the moment without doing more complex (and time consuming) changes in Ophan.

This change expands the scope of data recording to also include CCPA consent information. Because CCPA consent information has a different structure than TCF.v2, we needed to redesign the way the data is encoded to gracefully support both jurisdictions. 

Moreover we tried to also make it convenient to query that information once it ends up in the data lake, and notably to make it convenient to select the attribute we want to see without relying on an implicit ordering of the array elements. 

The chosen solution is simply to prefix the attributes with a unique code that can be used by SQL string functions. 

## Component Events (Updated)

The component events now looks as follows, respectively for TCF.v2 and CCPA

```
{
    "component": {
        "componentType": "CONSENT",
        "products": [],
        "labels": [
            "01:TCF.v2",
            "02:68916df3-d43e-45d1-a696-fade0d8879fd_4",
            "03:CPTHoTJPTHoTJAGABCENB-CgAP_AAG_AAAQ4IDtB9S7eTWMDeH5_(...) [reduced for convenience]"
        ]
    },
    "action": "MANAGE_CONSENT"
}
```

```
{
    "component": {
        "componentType": "CONSENT",
        "products": [],
        "labels": [
            "01:CCPA",
            "04:ba8edece-df47-433f-85d7-b5ae94e1d9a6",
            "05:false"
        ]
    },
    "action": "MANAGE_CONSENT"
}
```

The mapping between length 2 codes and their meaning is currently in the code, but will be promoted to T&C documentation.

```
Codes:
    01: Specify the jurisdiction. Accepted values: "TCF.v2" and "CCPA"
    02: TCF.v2 Consent UUID
    03: TCF.v2 Consent string
    04: CCPA (Consent) UUID
    05: Do not sell, boolean, with the serialization: true -> "true", false -> "false"

For TCF.v2, an exmaple of array is ["01:TCF.v2", "02:<consent UUID>"", "03:<consent string>""]
For CCPA,   an exmaple of array is ["01:CCPA", "04:<consent UUID>", "05:true"]
```

### CODE REVIEW

File: dotcom-rendering/src/web/browser/bootCmp/init.ts [hunk 1]
```diff
@@ -63,27 +63,78 @@ const init = async (): Promise<void> => {
...
+			we are going to use a specific convention.
+
+			Each element of the array represents a particular attribute. Note that the collection of attributes is different
+			from one jurisdiction to another. To have an unambiguous reading we prefix each attribute value with a code
+			that helps identify the attribute type.
+
+			An element of the array is then always of the form "<code>:<value>"
+
+			Codes:
+				01: Specify the jurisdiction. Accepted values: "TCF.v2" and "CCPA"
```
Comment (kenoir) [inline dotcom-rendering/src/web/browser/bootCmp/init.ts:93]: Can we use string identifiers that at least hint at the contained value? I think i'll be looking up what they mean whenever I come to look at this data. Definitely keep the documentation explaining the detail though.

A consistent string length for the ID is a good idea though as it helps with removing the prefix identifier when writing SQL.

Comment (shtukas) [reply dotcom-rendering/src/web/browser/bootCmp/init.ts:93]: Sure. Sure. Let me do that :)

Comment (kenoir) [reply dotcom-rendering/src/web/browser/bootCmp/init.ts:93]: Discussed with @shtukas, this is fine for now.

Comment (shtukas) [reply dotcom-rendering/src/web/browser/bootCmp/init.ts:93]: So... as discussed, let's try and run with the codes for the moment while we are in prototyping stage.

**Review by kenoir — APPROVED** [review]

File: dotcom-rendering/src/web/browser/bootCmp/init.ts [hunk 2]
```diff
@@ -63,27 +63,84 @@ const init = async (): Promise<void> => {
...
+			For CCPA,   an exmaple of array is ["01:CCPA", "04:<consent UUID>", "05:true"]
+
+			Note: it is possible to deprecate CODES, but they cannot the reused.
+		*/
+
 		const decideConsentString = () => {
 			if (consentState.tcfv2) {
 				return consentState.tcfv2?.tcString;
 			}
+			if (consentState.ccpa) {
```
Comment (DavidLawes) [inline dotcom-rendering/src/web/browser/bootCmp/init.ts:109]: I don't think we call `decideConsentString` for ccpa `consentState` (we assign 'true' or 'false' directly to `flag` on line 126)?

Regardless, I suppose it's good to explain how the consent string should be created for each different consent state, just thought I'd mention. Let me know if I've overlooked something!

**Review by DavidLawes — APPROVED** [review]
Lovely documentation 😄

File: dotcom-rendering/src/web/browser/bootCmp/init.ts (hunk 2 above)
Comment (shtukas) [reply dotcom-rendering/src/web/browser/bootCmp/init.ts:109]: Hi David, you are perfectly right. When I did the refactoring I didn't notice that `decideConsentString` was no longer called in the CCPA case. I actually got rid on that function. Thanks for noticing !

**Comment by shtukas** [comment]
Tested on CODE with UK IP Address

```
{"component":{"componentType":"CONSENT","products":[],"labels":["01:TCF.v2","02:ed84400e-e77f-42dc-9d2b-373d9a5be9e1_3","03:CPTHoTJPTHoTJAGABCENB-CgAP_AAG_AAAQ4IDtB9S7eTWMDeH5_Y7t0cYUX1R5_oeAijgEAF4IBwRKUIIwWEGAyJESIBqACEAYAIiJAIAdkGEAAAEAAYIABAAHIAEAEKAAAIAIEGAABAgIACAAIAAAAEAAQgAAUECAgiAAAcJYgSAAABAAAAAAAAAgEAAAAAgAAAAAAAAAAAAAAACgbRAhAAcABcAEIAOQAfgBkADQAG0ARwAkQBZgC5AHUAO6Ag4CEAERAJ2AT8ApYBbQC6wGAAYEAzIBrADXwHUAdUA7YB_wEPgJiAXaAxYBtAB4yAOAEwARwBHAEnAJiAWwAvMRAGAEMANkA1gB1QEOgMnEAAQASBIEwACwAKgAZAA5AB4AIAAZAA8gCIAIoATAAngBVADmAHoAPwAhABDACIAEcAJYATQApQBbgDZAHsAPgAfoBAwCOAEpANYAcQBDoCYgFsALzAYaAyQBk4QAEACQApYaAQAIYAbMBaAFpANYAdUBDoDGAGTioBAATABHAEcgLQAtICQQExALYAXmOgaAALAAqABkADkAHwAggBiAGQAPAAfQBEAEUAJgATwAqgBdAC-AGIAMwAcwA9AB-AEMAIgARwAlgBMACaAFGAKUAWIAtwBhADRAGyAPYAfoBAwCLAEcAJTAWgBaQC6gF5AOIAdQBDoCQQFsALtAXmAw0BjADJAGTjgAQAJABSEICwACwAMgAxACYAFUAL4AYgAzAB6AEcALEAYQBHACUwFoAWkA6gCQQEnALYAXaAycBwBAACAKQlAfAAWABkADgAHwAYgA8ACIAEwAKoAXwAxABmADaAIYARAAjgBRgClAFuAMIAbIBHIC0ALSAXUA6gCHQFsALtAXmAycBwBIACAKQpAoAAWABUADIAHIAPgBBADEAMgAeQBEAEUAJgATwApABVAC-AGIAMwAcwA_ACGAEQAKMAUoAsQBbgDCAGiANkAfoBFgCOAEpALyAh0BJwC2AF2gLzAYaAxgBkgDJygAIAEgApAAA.YAAAAAAAAAAA"]},"action":"MANAGE_CONSENT"}
```

**Comment by shtukas** [comment]
... and with a US IP address:


{"component":{"componentType":"CONSENT","products":[],"labels":["01:CCPA","04:7a95b064-c2ca-492e-beb1-3a3999aa72b9","05:false"]},"action":"MANAGE_CONSENT"}

**Pull request merged**